### PR TITLE
planner, session, util: support Insert hint

### LINKED ISSUE #16929: The memory_quota hint doesn't work in INSERT INTO statement
## Bug Report

Please answer these questions before submitting your issue. Thanks!
### 1. What did you do?

```
drop table if exists t;
create table t(a int, b int, c int);
insert into t values(1,1,1);
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
insert into t select * from t;
select /*+ memory_quota(1 MB) */ * from t;
insert into t select /*+ memory_quota(1 MB) */ * from t;
insert /*+ memory_quota(1 MB) */  into t select * from t;
```

### 2. What did you expect to see?
```
mysql> select /*+ memory_quota(1 MB) */ * from t;
ERROR 1105 (HY000): Out Of Memory Quota![conn_id=1]
mysql> insert into t select /*+ memory_quota(1 MB) */ * from t;
ERROR 1105 (HY000): Out Of Memory Quota![conn_id=1]
mysql> insert into /*+ memory_quota(1 MB) */  t select * from t;
ERROR 1105 (HY000): Out Of Memory Quota![conn_id=1]
```
### 3. What did you see instead?
```
mysql> select /*+ memory_quota(1 MB) */ * from t;
ERROR 1105 (HY000): Out Of Memory Quota![conn_id=1]
mysql> insert into t select /*+ memory_quota(1 MB) */ * from t;
Query OK, 65536 rows affected (0.67 sec)
Records: 65536  Duplicates: 0  Warnings: 0
mysql> insert into /*+ memory_quota(1 MB) */  t select * from t;
Query OK, 131072 rows affected (1.48 sec)
Records: 131072  Duplicates: 0  Warnings: 0
```
### 4. What version of TiDB are you using? (`tidb-server -V` or run `select tidb_version();` on TiDB)
```
Release Version: v4.0.0-rc-156-g27a1c2b0a-dirty
Git Commit Hash: 27a1c2b0a4cd53b0343a0d6fd6442f9a5be72fe5
Git Branch: release-4.0
UTC Build Time: 2020-04-29 06:12:12
GoVersion: go1.13.1
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
```

### PR DESCRIPTION
**Pull request by Yisaer**
### What problem does this PR solve?

Currently, the `Insert` didn't support any Optimizer Hints.  

Issue Number: close #16929 

###  What's Changed:

This request support `Insert /*+ OptimizerHint */  .... ` for InsertStmt.  For duplicated Supported Hints like `Insert /*+ OptimizerHint */ into t select  /*+ OptimizerHint */ ....` , we would send a warning message.

### Release note 
* Support `Insert` Optimizer Hints.

### CODE REVIEW

**Comment by Yisaer** [comment]
I want to add some unit tests for testing `insert/*+ memory_quota(1 byte) */  into t select * from t;` but don't know where I can add this test.

**Comment by Yisaer** [comment]
> nice work, it's almost done! could you add some tests?

I'd like to add some tests but I don't know where I add these case properly. (For example, which test file or test suite I can add the test) Could you give me some suggestions?

**Comment by SunRunAway** [comment]
> > nice work, it's almost done! could you add some tests?
> 
> I'd like to add some tests but I don't know where I add these case properly. (For example, which test file or test suite I can add the test) Could you give me some suggestions?

You can add tests [here](https://github.com/pingcap/tidb/blob/f72b5984dd33c5ca5e37bf812b8ecb7551fed0b0/session/session_test.go#L3071)

File: util/hint/hint_processor.go [hunk 1]
```diff
@@ -63,7 +63,12 @@ func ExtractTableHintsFromStmtNode(node ast.Node) []*ast.TableOptimizerHint {
 		return x.TableHints
 	case *ast.DeleteStmt:
 		return x.TableHints
-	// TODO: support hint for InsertStmt
+	case *ast.InsertStmt:
+		// If the Select of Insert have hints, it would replace the Insert hints
```
Comment (SunRunAway) [inline util/hint/hint_processor.go:67]: Does MySQL have this policy?

Comment (Yisaer) [reply util/hint/hint_processor.go:67]: In [Mysql 8.0 ducoument](https://dev.mysql.com/doc/refman/8.0/en/optimizer-hints.html)

> Duplicate hints: For a hint such as /*+ MRR(idx1) MRR(idx1) */, MySQL uses the first hint and issues a warning about the duplicate hint.

And I also test with mysql 8.0, It seems the hint in the select with a higher priority.

```sh
mysql> insert /*+ qb_name(qb1) */ into a select /*+ qb_name(qb2) */ 1;
Query OK, 1 row affected, 1 warning (0.08 sec)
Records: 1  Duplicates: 0  Warnings: 1

mysql> show warnings;
+---------+------+----------------------------------------------------------+
| Level   | Code | Message                                                  |
+---------+------+----------------------------------------------------------+
| Warning | 3126 | Hint QB_NAME(`qb1`) is ignored as conflicting/duplicated |
+---------+------+----------------------------------------------------------+
1 row in set (0.00 sec)

```

**Comment by Yisaer** [comment]
@SunRunAway @qw4990  PTAL when you are free.

File: util/hint/hint_processor.go (hunk 1 above)
Comment (SunRunAway) [reply util/hint/hint_processor.go:67]: But this change may result in unexpected behaviors. For example, w/ MySQL8.0
```mysql
mysql> insert into tt select /*+ max_execution_time(1000) */ sleep(3);
Query OK, 1 row affected, 1 warning (3.01 sec)
Records: 1  Duplicates: 0  Warnings: 1

mysql> show warnings;
+---------+------+-------------------------------------------------------------------------------------+
| Level   | Code | Message                                                                             |
+---------+------+-------------------------------------------------------------------------------------+
| Warning | 3125 | MAX_EXECUTION_TIME hint is supported by top-level standalone SELECT statements only |
+---------+------+-------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

File: go.sum [hunk 2]
```diff
@@ -4,22 +4,18 @@ cloud.google.com/go v0.38.0/go.mod h1:990N+gfupTy94rShfmMCWGDn0LpTmnzTp2qbd1dvSR
 cloud.google.com/go v0.44.1/go.mod h1:iSa0KzasP4Uvy3f1mN/7PiObzGgflwredwwASm/v6AU=
 cloud.google.com/go v0.44.2/go.mod h1:60680Gw3Yr4ikxnPRS/oxxkBccT6SA1yMk63TGekxKY=
 cloud.google.com/go v0.45.1/go.mod h1:RpBamKRgapWJb87xiFSdk4g1CME7QZg3uwTez+TSTjc=
-cloud.google.com/go v0.46.3 h1:AVXDdKsrtX33oR9fbCMu/+c1o8Ofjq6Ku/MInaLVg5Y=
```
Comment (SunRunAway) [inline go.sum:7]: Seems that `go.sum` should not be changed.

File: util/hint/hint_processor.go (hunk 1 above)
Comment (Yisaer) [reply util/hint/hint_processor.go:67]: You're right. What about directly return insert's hints here?

Comment (SunRunAway) [reply util/hint/hint_processor.go:67]: Yes, I think just let `qb_name(qb1)` affect with this SQL `insert /*+ qb_name(qb1) */ into a select /*+ qb_name(qb2) */ 1;` is ok.

**Comment by SunRunAway** [comment]
@Yisaer Please resolve [this comment](https://github.com/pingcap/tidb/pull/16966#discussion_r431856880)

File: util/hint/hint_processor.go (hunk 1 above)
Comment (Yisaer) [reply util/hint/hint_processor.go]: updated.

File: session/session_test.go [hunk 3]
```diff
@@ -3091,6 +3091,16 @@ func (s *testSessionSuite2) TestStmtHints(c *C) {
 	c.Assert(tk.Se.GetSessionVars().StmtCtx.MemTracker.CheckBytesLimit(val), IsTrue)
 	c.Assert(tk.Se.GetSessionVars().StmtCtx.GetWarnings()[0].Err.Error(), Equals, "Setting the MEMORY_QUOTA to 0 means no memory limit")
 
+	tk.MustExec("use test")
+	tk.MustExec("create table t1(a int);")
+	tk.MustExec("insert /*+ MEMORY_QUOTA(1 MB) */ into t1 (a) values (1);")
+	val = int64(1) * 1024 * 1024
+	c.Assert(tk.Se.GetSessionVars().StmtCtx.MemTracker.CheckBytesLimit(val), IsTrue)
+
+	tk.MustExec("insert /*+ MEMORY_QUOTA(1 MB) */  into t1 select /*+ MEMORY_QUOTA(3 MB) */ * from t1;")
```
Comment (SunRunAway) [inline session/session_test.go:3100]: Should append a warning message because of `/*+ MEMORY_QUOTA(3 MB) */`

Comment (Yisaer) [reply session/session_test.go:3100]: updated.

**Comment by zz-jason** [comment]
@Yisaer please follow the PR title format:

```
<!-- Thank you for contributing to TiDB!

PR Title Format:
1. pkg [, pkg2, pkg3]: what's changed
2. *: what's changed

-->
```

**Comment by Yisaer** [comment]
updated.

File: util/hint/hint_processor.go [hunk 4]
```diff
@@ -28,6 +28,13 @@ import (
 	"go.uber.org/zap"
 )
 
+var supportedHintNameForInsertStmt = map[string]struct{}{}
+
+func init() {
+	supportedHintNameForInsertStmt["MEMORY_QUOTA"] = struct{}{}
+	supportedHintNameForInsertStmt["memory_quota"] = struct{}{}
```
Comment (zz-jason) [inline util/hint/hint_processor.go:35]: Please maintain only one map entry for one hint. BTW, can we support all hints for `INSERT` statements?

Comment (Yisaer) [reply util/hint/hint_processor.go:35]: updated.  And for this question `can we support all hints for INSERT statements` , I suggest discussing this in another issue.

File: util/hint/hint_processor.go [hunk 5]
```diff
@@ -55,22 +61,57 @@ func (hs *HintsSet) ContainTableHint(hint string) bool {
...
+			if supportedHint != nil {
+				var duplicatedHint *ast.TableOptimizerHint
+				for _, hint := range ExtractTableHintsFromStmtNode(x.Select, nil) {
+					if hint.HintName.L == supportedHint.HintName.L {
+						duplicatedHint = hint
+						break
+					}
+				}
+				if duplicatedHint != nil {
+					warning := fmt.Sprintf("Hint %s(`%v`) is ignored as conflicting/duplicated", duplicatedHint.HintName.O, duplicatedHint.HintData)
```
Comment (SunRunAway) [inline util/hint/hint_processor.go:105]: https://dev.mysql.com/doc/refman/8.0/en/server-error-reference.html#error_er_warn_conflicting_hint
It's better to define this error message in `tidb/errno` package, and use an error variable here.

Comment (Yisaer) [reply util/hint/hint_processor.go]: updated

File: util/hint/hint_processor.go (hunk 4 above)
Comment (zz-jason) [reply util/hint/hint_processor.go:36]: which issue?

Comment (Yisaer) [reply util/hint/hint_processor.go:36]: https://github.com/pingcap/tidb/issues/18049 I think we can discuss the ranges of hints we are going to support for `Insert` in this issue. @zz-jason

**Comment by SunRunAway** [comment]
@winoros @eurekaka  Please help merge.

**Comment by XuHuaiyu** [comment]
@SunRunAway 
Should we cherry-pick this to release-4.0?

**Pull request merged**